feature: add docs to composite credential (#111)
UnumID/UnumID.github.io: docs/api-overview.mdx

1. User's email or phone
2. (Optional) Credential request(s). _Required if Partner UUID is not provided._
3. (Optional) Partner UUID. _More information on the partner UUID use case can be found in the [Issuance Guide](/issuance-guide#already-issued-credentials)._
4. (Optional) List of the sub-credentials that make up the credential. _If not provided, all sub-credentials will be returned._

A credential requests encodes which <Tip type="credential">credentials</Tip> you're asking the user to share with your <Tip type="brand"/>. It's a list of one or more [`CredentialRequest`](https://docs.verified.inc/types/interfaces/credentialrequest.html) objects.

<CredentialRequestCodeSnippet />

If you list multiple <Tip type="issuer">issuers</Tip>, the user can include a credential issued by _any one_ of those listed. If `issuers` is left empty the user can include a credential with matching type issued by _any_ issuer.

> :::note Note

Reviewer: ```suggestion
:::note
```

The Credential <Tip type="request">requests</Tip> attribute is required if a Partner UUID is **not** provided. If Partner UUID is provided, the Credential requests attribute is not used. Instead, a pre-defined set of accepting partner's credentials requests are used.
_More information on the partner UUID use case can be found in the [Issuance Guide](/issuance-guide#already-issued-credentials)._
:::

> :::note Note

Reviewer: ```suggestion
:::tip
```

If <b>mandatory</b> field are missing, then the credential will be treated as optional.
:::

> :::note Note

Reviewer: ```suggestion
:::note
```

If <b>mandatory</b> or <b>required</b> fields are both missing, then the credential will be treated as optional.
:::

> :::note Note 3

Reviewer: Mind dropping this number Note convention? I think we could choose different syntax highlights between the three "notes" to break up the sameness.

Reviewer: Also, please mention that `required` in Note 2 is now deprecated.

The `url` from the response ought to then be enriched with a `redirectUrl` query param that will take the next step in your target service's user onboarding webpage. _More information on this can be found in the [Acceptance Guide](/acceptance-guide)._

> :::info Example Credential Request

Reviewer: I very much like this example section but we usually have left detailed examples for the guides... I suppose we may want to provide detailed example request bodies for all the api endpoints in this manner if we are doing for /hasMatchingCreds.

Author: Yeap, we can do this one day

> So for example, if you want to check if a user has credentials of type `SsnCredential`, `FullNameCredential` and `AddressCredential` issued by any issuer you would send the following request:

Reviewer: ```suggestion
For example, if you want to check if a user has credentials of type `SsnCredential`, `FullNameCredential` and `AddressCredential` issued by any issuer you would send the following request:
```

<details>
<summary>Request Body</summary>
<p>

```json
{
  "email": "richard@piedpiper.net",
  "credentialRequests": [
    {
      "type": "SsnCredential",
      "issuers": [],
      "mandatory": "yes"
    },
    {
      "type": "FullNameCredential",
      "issuers": [],
      "mandatory": "yes"
    },
    {
      "type": "AddressCredential",
      "issuers": [],
      "children": [
        {
          "type": "CityCredential",
          "issuers": [],
          "mandatory": "if_available"
        },
        {
          "type": "StateCredential",
          "issuers": [],
          "mandatory": "yes"
        },
        {
          "type": "CountryCredential",
          "issuers": [],
          "mandatory": "yes"
        },
        {
          "type": "ZipCodeCredential",
          "issuers": []
        }
      ]
    }
  ]
}
```

</p>
</details>

> In the given example, the user will be asked to share their `SsnCredential`, `FullNameCredential` and `AddressCredential` credentials.

Author: But do u think we need remove all that ? or just rewritte ?

> In the given example, the user will be asked to share their `SsnCredential`, `FullNameCredential` and `AddressCredential` credentials.

Reviewer: I think it should be removed from this section. Further down in the example, you call out the result of this request body. I think stating here is confusing (before explaining the request body). Also, the section collapsed it read a little funny.

> Let's breakdown to understand what is happening here:

Reviewer 2: ```suggestion
Let's break this down to understand what is happening here:
```

- `SsnCredential` is a "[single attribute atomic credential](/schema#schema-definitions)" and therefore does not have any sub-credentials. Because it is marked as `mandatory`, it only matches if the user has a credential.

```json title="SnnCredential"
{
  "type": "SsnCredential",
  "issuers": [],
  "mandatory": "yes"
}
```

- `FullNameCredential` is a "[composite credential](/schema#schema-definitions)" and therefore has *2 sub-credentials*¹: `FirstNameCredential`, `LastNameCredential`. Because it is marked as `mandatory`, and it has no `children` attribute, it only matches if the user has a credential with all sub-credentials.

```json title="FullNameCredential"
{
  "type": "FullNameCredential",
  "issuers": [],
  "mandatory": "yes"
}
```

> > _¹The `FullNameCredential` also has a `MiddleNameCredential`, but it's `optional`._

Reviewer: Hmm this is confusing even for me! Are we saying that FirstName and LastName are mandatory but MiddleName is not if the FullName composite is? 

Shouldn't MIddleName be mandatory just like First and Last?

Author: I agree with Jacob

Reviewer: But why then is FirstName and LastName implicitly considered manadatory but MiddleName is not? Where is this defined? 

I don't like it because how is one suppose to know what is mandatory by default in a composite cred and not? I think all the atomic creds in a composite cred should be considered mandatory or not mandatory.

Reviewer 3: It definitely needs to be clear (at least in documentation, ideally encoded in the schema somehow) which child credentials are mandatory and which are not.

Author: I thought the `Schema Library` section was for that

https://github.com/UnumID/UnumID.github.io/pull/111#discussion_r1295311046

- The `AddressCredential` credential is made up of 6 sub-credentials: `Line1Credential`, `Line2Credential`, `CityCredential`, `StateCredential`, `CountryCredential` and `ZipCodeCredential`.

  But because we specified a `children` attribute, we are only asking for the `CityCredential`, `StateCredential`, `CountryCredential` and `ZipCodeCredential` sub-credentials. So `Line1Credential` and `Line2Credential` are ignored.

  ```json title="AddressCredential"
  {
    "type": "AddressCredential",
    "issuers": [],
    "children": [
      {
        "type": "CityCredential",
        "issuers": [],
        "mandatory": "if_available"
      },
      {
        "type": "StateCredential",
        "issuers": [],
        "mandatory": "yes"
      },
      {
        "type": "CountryCredential",
        "issuers": [],
        "mandatory": "yes"
      },
      {
        "type": "ZipCodeCredential",
        "issuers": []
      }
    ]
  }
  ```

>   The user will be asked to share the `CityCredential` if available (this mean if the user has a credential of type `CityCredential` it will mandory be shared, but if the user does not have a credential of type `CityCredential` the user will still be able to share the `AddressCredential` credential).

Reviewer 2: ```suggestion
The user will be asked to share the `CityCredential` if available (this means if the user has a credential of type `CityCredential` it will be mandatory, but if the user does not have a credential of type `CityCredential` the user will still be able to share the `AddressCredential` credential).
```

Credential data is stored securely via a data privacy vault that encrypts and tokenizes the data. The data is only retrieved and decrypted for the brand that the <Tip type="user"/> agrees to share with.
:::

:::info Composite Credentials

> Composite credentials are credentials that are made up of other credentials. For example, an `FullNameCredential` is made up of `FirstNameCredential`, `MiddleNameCredential`(optional) and `LastNameCredential`. When issuing a composite credential, you must provide the child credentials as well.

Reviewer: I very much think that just having a random rule that MiddleName is optional does not make sense. Are there any other random implicit rules like this for other composite creds? This subjective optional should be avoided. Explicit definitions are much more desirable.

Author: Line2 it's another example of optional data

Reviewer: Hmm =/. How is this documented? How is one supposed to know this? I don't think it is in our json schemas?

Author: <img width="615" alt="image" src="https://github.com/UnumID/UnumID.github.io/assets/11665444/7812deab-0172-4f19-9d26-fd5490973130">

In our JSON schema we have required field when is required

Reviewer 3: I agree that we should be explicit about which child credentials are required/not required when issuing a composite credential. I'm sure there are a lot more cases like Address, where obviously line2 can't be required because many addresses don't have it, or FullName, where many issuers probably don't collect middle name. Especially this early on we are going to need to be flexible in allowing issuers to issue what they have and not try to force unnecessary requirements on them.

Reviewer: Thanks for providing the example Leo. As long as this is explicitly defined in the json schema I suppose I'm ok with it.

> Composite credentials are credentials that are made up of other credentials. For example, an `FullNameCredential` is made up of `FirstNameCredential`, `MiddleNameCredential`(optional) and `LastNameCredential`. When issuing a composite credential, you must provide the child credentials as well.

Reviewer 2: ```suggestion
Composite credentials are credentials that are made up of other credentials. For example, a `FullNameCredential` is made up of `FirstNameCredential`, `MiddleNameCredential`(optional), and `LastNameCredential`. When issuing a composite credential, you must provide the child credentials as well.
```

<details>
<summary>Example</summary>

```typescript title="Composite Credential"
{
  "type": "FullNameCredential",
  // based on the properties attribute of the schema
  // highlight-start
  // note that the data is an array of objects, not a single object
  "data": [
    {
      "type": "FirstNameCredential",
      "data": {
        "firstName": "Richard"
      }
    },
    {
      "type": "LastNameCredential",
      "data": {
        "lastName": "Hendricks"
      }
    }
  ]
  // highlight-end
}
```

</details>

> See the [schema](/schema) for more information.

Reviewer: ```suggestion
See [Data Schemas](/schema) for more information.
```
